game-app/game-core/src/main/java/games/strategy/triplea/delegate/data/CasualtyDetails.java

```diff
@@ -76,13 +78,41 @@ private CasualtyDetails ensureUnitsAreTakenFirst(
             getKilled().stream()
                 .filter(matcher)
                 .collect(Collectors.groupingBy(UnitOwner::new, Collectors.toList())));
-    final List<Unit> damaged =
-        ensureUnitsAreTakenFirst(
-            comparator,
-            targetsGroupedByOwnerAndType,
-            getDamaged().stream()
-                .filter(matcher)
-                .collect(Collectors.groupingBy(UnitOwner::new, Collectors.toList())));
+
+    final List<Unit> damaged = new ArrayList<>();
+
+    final Map<UnitOwner, List<Unit>> oldTargetAirUnitsToTakeHits =
+        getDamaged().stream()
+            .filter(matcher)
+            .collect(Collectors.groupingBy(UnitOwner::new, Collectors.toList()))
+        ;
+
+    for(final Map.Entry<UnitOwner, List<Unit>> entry : oldTargetAirUnitsToTakeHits.entrySet()) {
+      final UnitAttachment ua = UnitAttachment.get(entry.getKey().getType());
+      final int hitPointsOfType = ua.getHitPoints();
+
+      final List<Unit> allTargetUnitsOfOwnerAndTypeThatCanTakeHits = new ArrayList<>(
+          targetsGroupedByOwnerAndType.get(entry.getKey()));
+
+      allTargetUnitsOfOwnerAndTypeThatCanTakeHits.sort(comparator.reversed());
+
+      final int hits = entry.getValue().size();
+
+      int hitsTaken = 0;
+      for (final Unit unit : allTargetUnitsOfOwnerAndTypeThatCanTakeHits) {
+        // Stop if we have already selected as many hits as there are targets
+        if (hitsTaken >= hits) {
+          break;
+        }
+
+        for (int hitPointsUnitCanTakeWithoutBeingKilled = hitPointsOfType - (1 + unit.getHits());
```

Comment:
A quick comment on this for loop could help, it's a bit difficult to follow. Extracting some of the values to named variables might also help the readability.

Reply:
I had used a similar loop in the existing code as the basis for my coding.
But I agree with you. I have rewirtten the loop completely.